File: public/python/054-coin-change.py
# LeetCode Problem #322: Coin Change
# Difficulty: Medium
# Link: https://leetcode.com/problems/coin-change/

from typing import List
from collections import deque
from functools import lru_cache

# ─────────────────────────────────────────────
# APPROACH 1: Recursive Memoization | O(S·n) time | O(S) space
# EXPLAIN: Top-down recursion tries every coin at each remaining amount; cache avoids recomputation.
# WHEN: Good starting point — models the problem naturally as a decision tree.

def coinChange_memo(coins: List[int], amount: int) -> int:
    @lru_cache(maxsize=None)
    def dp(rem):
        if rem == 0:
            return 0
        if rem < 0:
            return float('inf')
        return min(1 + dp(rem - c) for c in coins)

    result = dp(amount)
    return result if result != float('inf') else -1


# APPROACH 2: BFS (Level-by-Level) | O(S·n) time | O(S) space
# EXPLAIN: BFS treats amount as target; each level adds one coin; first time we reach 0 is fewest coins.
# WHEN: When you think of it as a shortest-path problem on an implicit graph.

def coinChange_bfs(coins: List[int], amount: int) -> int:
    if amount == 0:
        return 0
    visited = {amount}
    queue = deque([amount])
    steps = 0
    while queue:
        steps += 1
        for _ in range(len(queue)):
            rem = queue.popleft()
            for c in coins:
                nxt = rem - c
                if nxt == 0:
                    return steps
                if nxt > 0 and nxt not in visited:
                    visited.add(nxt)
                    queue.append(nxt)
    return -1


# APPROACH 3: Bottom-Up DP | O(S·n) time | O(S) space
# EXPLAIN: dp[i] = minimum coins to make amount i; build from 0 up to amount.
# WHEN: Best practical choice — simple loop, no recursion overhead.

def coinChange_dp(coins: List[int], amount: int) -> int:
    dp = [float('inf')] * (amount + 1)
    dp[0] = 0
    for i in range(1, amount + 1):
        for c in coins:
            if c <= i:
                dp[i] = min(dp[i], dp[i - c] + 1)
    return dp[amount] if dp[amount] != float('inf') else -1


# APPROACH 4: Bottom-Up DP with Sorted Early-Exit | O(S·n) time | O(S) space
# EXPLAIN: Sort coins ascending then break inner loop when coin > remaining amount.
# WHEN: Minor practical speedup when coins contain large denominations.

def coinChange_sorted(coins: List[int], amount: int) -> int:
    sorted_coins = sorted(coins)
    dp = [float('inf')] * (amount + 1)
    dp[0] = 0
    for i in range(1, amount + 1):
        for c in sorted_coins:
            if c > i:
                break
            dp[i] = min(dp[i], dp[i - c] + 1)
    return dp[amount] if dp[amount] != float('inf') else -1


# ── quick tests ──────────────────────────────────────────────────────────────
if __name__ == '__main__':
    cases = [
        ([1, 5, 10, 25], 30, 2),
        ([1, 2, 5], 11, 3),
        ([2], 3, -1),
        ([1], 0, 0),
        ([1], 1, 1),
        ([186, 419, 83, 408], 6249, 20),
    ]
    for fn in (coinChange_memo, coinChange_bfs, coinChange_dp, coinChange_sorted):
        for coins, amount, expected in cases:
            result = fn(coins, amount)
            assert result == expected, (
                f'{fn.__name__}({coins}, {amount}) = {result}, expected {expected}'
            )
    print('All tests passed.')

# Made with Bob

File: public/java/054-coin-change.java
/*
 * LeetCode Problem #322: Coin Change
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/coin-change/
 */
import java.util.*;

class Solution {

    // APPROACH 1: Recursive Memoization (Top-Down DP) | O(S*n) time | O(S) space
    // EXPLAIN: For each remaining amount S, try every coin and cache the min-coin result.
    // WHEN: Use when top-down recursion is clearer; S = target amount, n = coins count.

    public int coinChange_memo(int[] coins, int amount) {
        int[] memo = new int[amount + 1];
        Arrays.fill(memo, -1);
        int result = dfs(coins, amount, memo);
        return result == Integer.MAX_VALUE ? -1 : result;
    }

    private int dfs(int[] coins, int rem, int[] memo) {
        if (rem == 0) return 0;
        if (rem < 0) return Integer.MAX_VALUE;
        if (memo[rem] != -1) return memo[rem];
        int best = Integer.MAX_VALUE;
        for (int c : coins) {
            int sub = dfs(coins, rem - c, memo);
            if (sub != Integer.MAX_VALUE)
                best = Math.min(best, sub + 1);
        }
        memo[rem] = best;
        return best;
    }

    // APPROACH 2: BFS (Shortest Path) | O(S*n) time | O(S) space
    // EXPLAIN: Treat coin change as BFS from 0 to amount; first time we reach amount is fewest coins.
    // WHEN: Use when you think of it as a shortest-path problem on an implicit graph.

    public int coinChange_bfs(int[] coins, int amount) {
        if (amount == 0) return 0;
        boolean[] visited = new boolean[amount + 1];
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(0);
        visited[0] = true;
        int level = 0;
        while (!queue.isEmpty()) {
            level++;
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int cur = queue.poll();
                for (int c : coins) {
                    int next = cur + c;
                    if (next == amount) return level;
                    if (next < amount && !visited[next]) {
                        visited[next] = true;
                        queue.offer(next);
                    }
                }
            }
        }
        return -1;
    }

    // APPROACH 3: Bottom-Up DP | O(S*n) time | O(S) space
    // EXPLAIN: dp[i] = fewest coins to make amount i; fill from 1 to amount using each coin.
    // WHEN: Use as the canonical interview answer — avoids recursion stack overhead.

    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1);      // sentinel "infinity"
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int c : coins) {
                if (c <= i) {
                    dp[i] = Math.min(dp[i], dp[i - c] + 1);
                }
            }
        }
        return dp[amount] > amount ? -1 : dp[amount];
    }

    // APPROACH 4: Bottom-Up DP with Sorted Early-Exit | O(S*n) time | O(S) space
    // EXPLAIN: Same as Approach 3 but sort coins ascending and break when coin > remaining amount.
    // WHEN: Minor optimisation when coins have large denominators — avoids unnecessary iterations.

    public int coinChange_sorted(int[] coins, int amount) {
        Arrays.sort(coins);
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int c : coins) {
                if (c > i) break;
                dp[i] = Math.min(dp[i], dp[i - c] + 1);
            }
        }
        return dp[amount] > amount ? -1 : dp[amount];
    }
}

// Made with Bob
